tensorflow/java/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -34,6 +41,115 @@
+   * or {@link LongBuffer} to create a Tensor with a datatype of {@code FLOAT}, {@code INT32}, {@code DOUBLE},
+   * or {@code INT64} respectively.
+   *
+   * <p> This method copies
+   * <i>n</i>&nbsp;=&nbsp;<tt>src.remaining()</tt> elements from the given
+   * buffer into the tensor, starting at the buffer's current position.
+   * The position of the buffer is then incremented by <i>n</i>.
+   *
+   * @param dataType the tensor datatype.
+   * @param shape the tensor shape.
+   * @param data a buffer containing the tensor data.
+   */
+  public static Tensor create(DataType dataType, long[] shape, Buffer data) {
+    Tensor t = new Tensor();
+    t.dtype = dataType;
+    t.shapeCopy = Arrays.copyOf(shape, shape.length);
+
+    int elemsize = 0;
+    if(data instanceof ByteBuffer) elemsize = 1;
+    else if(data instanceof IntBuffer) elemsize = Integer.SIZE / Byte.SIZE;
+    else if(data instanceof FloatBuffer) elemsize = Float.SIZE / Byte.SIZE;
+    else if(data instanceof DoubleBuffer) elemsize = Double.SIZE / Byte.SIZE;
+    else if(data instanceof LongBuffer) elemsize = Long.SIZE / Byte.SIZE;
+    else throwIncompatibleBuffer();
+
+    t.nativeHandle = allocate(t.dtype.c(), t.shapeCopy, elemsize * data.remaining());
+    try {
+      ByteBuffer dst = getBuffer(t.nativeHandle);
+      if(data instanceof ByteBuffer) {
+        dst.put((ByteBuffer) data);
+      }
+      else if(data instanceof IntBuffer) {
+        if(t.dtype != DataType.INT32) throwIncompatibleBuffer();
+        dst.asIntBuffer().put((IntBuffer) data);
+      }
+      else if(data instanceof FloatBuffer) {
+        if(t.dtype != DataType.FLOAT) throwIncompatibleBuffer();
+        dst.asFloatBuffer().put((FloatBuffer) data);
+      }
+      else if(data instanceof DoubleBuffer) {
+        if(t.dtype != DataType.DOUBLE) throwIncompatibleBuffer();
+        dst.asDoubleBuffer().put((DoubleBuffer) data);
+      }
+      else if(data instanceof LongBuffer) {
+        if(t.dtype != DataType.INT64) throwIncompatibleBuffer();
+        dst.asLongBuffer().put((LongBuffer) data);
+      }
+
+      return t;
+    }
+    catch(RuntimeException e) {
+      delete(t.nativeHandle);
+      throw e;
+    }
+  }
+
+  /**
+   * Gets the size (in bytes) of the tensor data.
+   *
+   * <p>Use this size information when constructing a buffer with which to call {@link #readData(Buffer)}.
+   */
+  public int getDataByteSize() {
+    return getBuffer(nativeHandle).remaining();
+  }
+    
+  /**
+   * Reads the tensor data into the given buffer.
+   *
+   * <p>Provide a {@link ByteBuffer} to read a Tensor of any datatype.  Note that primitive
+   * data is in native byte order. Provide a {@link FloatBuffer}, {@link IntBuffer}, {@link DoubleBuffer},
+   * or {@link LongBuffer} to read a Tensor with a datatype of {@code FLOAT}, {@code INT32}, {@code DOUBLE},
+   * or {@code INT64} respectively.
+   *
+   * @param dst the destination buffer.
+   */
+  public void readData(Buffer dst) {
```

Comment:
Keep in mind that `FloatBuffer`s that wrap an array are writable, so `copyTo` applied to a float buffer is more efficient than it would be if applied to a `ByteBuffer` (assuming you ultimately want a `float[]`).

I agree that the lack of guidance on sizing of typed buffers is inconvenient, but some utility methods might be enough.

Keeping with the power-over-convenience assumption, I'm inclined to provide optimized transfers.